sklearn/metrics/tests/test_common.py

```diff
@@ -1742,55 +1742,50 @@ def check_array_api_metric(
     metric_np = metric(y_true_np, y_pred_np, sample_weight=sample_weight)
 
     with config_context(array_api_dispatch=True):
-        if sample_weight is not None:
-            sample_weight = xp.asarray(sample_weight, device=device)
+        sample_weight = xp.asarray(sample_weight, device=device)
         metric_xp = metric(y_true_xp, y_pred_xp, sample_weight=sample_weight)
 
         assert_allclose(
             metric_xp,
             metric_np,
-            atol=_atol_for_type(dtype),
+            atol=_atol_for_type(dtype_name),
         )
 
 
 def check_array_api_binary_classification_metric(
-    metric, array_namespace, device, dtype
+    metric, array_namespace, device, dtype_name
 ):
     y_true_np = np.array([0, 0, 1, 1])
     y_pred_np = np.array([0, 1, 0, 1])
+    sample_weight = np.array([0.0, 0.1, 2.0, 1.0], dtype=dtype_name)
+
     check_array_api_metric(
-        metric, array_namespace, device, dtype, y_true_np=y_true_np, y_pred_np=y_pred_np
+        metric,
+        array_namespace,
+        device,
+        dtype_name,
+        y_true_np=y_true_np,
+        y_pred_np=y_pred_np,
+        sample_weight=sample_weight,
     )
-    if "sample_weight" in signature(metric).parameters:
-        check_array_api_metric(
-            metric,
-            array_namespace,
-            device,
-            dtype,
-            y_true_np=y_true_np,
-            y_pred_np=y_pred_np,
-            sample_weight=np.array([0.0, 0.1, 2.0, 1.0]),
-        )
```

Comment:
For metrics that supports `sample_weight` we want to test both `sample_weith=None` and `sample_weight=sample_weight`.

Reply:
So I had to re-write the metric a bit because the previous version never could have worked with metric that do not expose a `sample_weight` argument. Not very satisfied with how it looks, but it's done.

Reply:
Actually, even with the current state of your PR, we never call `check_array_api_metric(..., sample_weight=_NoSampleWeight)`, so the problem would still be there.

So if there is a metric that does not accept `sample_weight` that will crash.

I did a quick review of the API doc, and I could not find any such metrics for classification or regression. So this is a YAGNI. Let's keep the original, simpler definition of `check_array_api_metric` and just remove the `if "sample_weight" in signature(metric).parameters` test.

The code will be simpler, and failures will be easier to inspect from the traceback without having to reason about without the for loop at line 1771.